netlogo-headless/test/benchdumps/VirusNet.txt

```diff
@@ -1165,70 +1165,43 @@ procedure SETUP-SPATIALLY-CLUSTERED-NETWORK:[NUM-LINKS _repeatlocal:1]{OTPL}:
                L8
                FRAME SAME1 org/nlogo/agent/AgentSet
                 ASTORE 4
-                GETSTATIC org/nlogo/agent/LinkManager$.MODULE$ : Lorg/nlogo/agent/LinkManager$;
-                ALOAD 4
-                INVOKEVIRTUAL org/nlogo/agent/LinkManager$.mustNotBeDirected (Lorg/nlogo/agent/AgentSet;)Lscala/Option;
-                ASTORE 5
-                ALOAD 5
-                INVOKEVIRTUAL scala/Option.isDefined ()Z
-                IFEQ L9
-                NEW org/nlogo/nvm/RuntimePrimitiveException
-                DUP
-                ALOAD 1
-                ALOAD 0
-                ALOAD 5
-                INVOKEVIRTUAL scala/Option.get ()Ljava/lang/Object;
-                CHECKCAST java/lang/String
-                INVOKESPECIAL org/nlogo/nvm/RuntimePrimitiveException.<init> (Lorg/nlogo/api/Context;Lorg/nlogo/nvm/Instruction;Ljava/lang/String;)V
-                ATHROW
-               L9
-               FRAME APPEND [org/nlogo/agent/AgentSet scala/Option]
                 ALOAD 0
                 GETFIELD org/nlogo/prim/_asm_proceduresetupspatiallyclusterednetwork_not_5.world : Lorg/nlogo/agent/World;
                 INVOKEVIRTUAL org/nlogo/agent/World.linkManager ()Lorg/nlogo/agent/LinkManager;
+                ASTORE 5
+                GETSTATIC scala/Predef$.MODULE$ : Lscala/Predef$;
+                ALOAD 5
                 ALOAD 3
                 ALOAD 2
                 ALOAD 4
-                ICONST_1
-                INVOKEINTERFACE org/nlogo/agent/LinkManager.findLinkEitherWay (Lorg/nlogo/agent/Turtle;Lorg/nlogo/agent/Turtle;Lorg/nlogo/agent/AgentSet;Z)Lorg/nlogo/agent/Link;
-                ASTORE 6
-                ALOAD 6
-                IFNULL L10
-                ALOAD 6
-                INVOKEVIRTUAL org/nlogo/agent/Link.isDirectedLink ()Z
-                IFEQ L11
-               L10
-               FRAME APPEND [org/nlogo/agent/Link]
-                ICONST_0
-                GOTO L12
-               L11
-               FRAME SAME
-                ICONST_1
-               L12
-               FRAME SAME1 I
+                INVOKEINTERFACE org/nlogo/agent/LinkManager.linksWith (Lorg/nlogo/agent/Turtle;Lorg/nlogo/agent/Turtle;Lorg/nlogo/agent/AgentSet;)[Lorg/nlogo/agent/Link;
+                CHECKCAST [Ljava/lang/Object;
+                INVOKEVIRTUAL scala/Predef$.refArrayOps ([Ljava/lang/Object;)Lscala/collection/mutable/ArrayOps;
+                INVOKEINTERFACE scala/collection/mutable/ArrayOps.nonEmpty ()Z
```

Comment:
`nonEmpty` is definitely more descriptive, but on a pure-performance level I suspect using `.length > 0` would be slightly faster (field access and compare vs loading object `scala.Predef$` and invoking method). Is performance of `link-neighbor` important in most link-based models? This bytecode is generated by `_linkneighbor.scala` and is generated by the following line of the VirusNet Benchmark:

```
let choice (min-one-of (other turtles with [not link-neighbor? myself]) [distance myself])
```